src/const_permutation.rs

```diff
@@ -0,0 +1,440 @@
+use core::fmt::Debug;
+
+use std::ops;
+
+#[cfg(feature = "random")]
+use rand::Rng;
+
+use crate::util::is_permutation;
+#[cfg(feature = "random")]
+use crate::Permutations;
+use crate::TryFromError;
+
+/// A permutation with a constant length fixed at compile time.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ConstPermutation<const N: usize>([usize; N]);
+impl<const N: usize> ConstPermutation<N> {
+    /// Returns the identity permutation.
+    pub fn identity() -> Self {
+        ConstPermutation((0..N).collect::<Vec<_>>().try_into().unwrap())
+    }
+    /// Returns the permutation which rotates r steps to the left.
+    pub fn rotation_left(r: usize) -> Self {
+        ConstPermutation(
+            (0..N)
+                .map(|i| (i + r) % N)
+                .collect::<Vec<_>>()
+                .try_into()
+                .unwrap(),
+        )
+    }
+    /// Returns the permutation which rotates r steps to the right.
+    pub fn rotation_right(r: usize) -> Self {
+        ConstPermutation::rotation_left(N - (r % N))
+    }
+    /// Returns the permutation which exchanges the elements at i and j.
+    pub fn transposition(i: usize, j: usize) -> Self {
+        assert!(i < N && j < N);
+        ConstPermutation(
+            (0..N)
+                .map(|k| {
+                    if k == i {
+                        j
+                    } else if k == j {
+                        i
+                    } else {
+                        k
+                    }
+                })
+                .collect::<Vec<_>>()
+                .try_into()
+                .unwrap(),
+        )
+    }
+
+    /// Returns a random permutation.
+    ///
+    /// Uses a uniform distribution.
+    #[cfg(feature = "random")]
+    pub fn random<R>(rng: &mut R) -> Self
+    where
+        R: Rng,
+    {
+        let ps = Permutations::new(N);
+        let i = rng.gen_range(0..ps.len());
+        ps.get(i).expect("random index out of range")
+    }
+    /// Applies the permutation to an element.
+    pub fn apply(&self, i: usize) -> usize {
+        self.0[i]
+    }
+    /// Returns an array permuted by this permutation.
+    pub fn permute<T>(&self, a: &[T; N]) -> [T; N]
+    where
+        T: Clone + Debug,
+    {
+        (0..N)
+            .map(|i| a[self.apply(i)].clone())
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap()
+    }
+    /// Returns the composition of the permutation with itself.
+    pub fn square(&self) -> Self {
+        self * self
+    }
+    /// Returns the composition of the permutation with itself `exp` number of times.
+    pub fn pow(&self, exp: u32) -> Self {
+        if exp == 0 {
+            ConstPermutation::identity()
+        } else if exp == 1 {
+            self.clone()
+        } else if exp % 2 == 0 {
+            self.square().pow(exp / 2)
+        } else {
+            self * self.pow(exp - 1)
+        }
```

Comment:
What would you think of implementing `MulAssign` for `ConstPermutation`? Then I think this could be written iteratively using exponentiation by squaring to avoid creating multiple intermediate arrays&#8202;&mdash;&#8202;which, since they're stored on the stack, could result in a stack overflow much more quickly than with `Box<[usize]>` in `Permutation`.

```suggestion
        let mut result = Self::identity();
        let mut term = self.clone();
        
        while exp > 0 {
            if exp % 2 == 1 {
                result *= term;
            }
            term *= term;
            exp /= 2;
        }
        
        return result;
```

Reply:
And for what it's worth, `MulAssign` could be implemented space-efficiently using `permute_in_place` from #3.

```rust
impl<const N: usize> ops::MulAssign<ConstPermutation<N>> for ConstPermutation<N> {
    fn mul_assign(&mut self, rhs: ConstPermutation<N>) {
        rhs.permute_in_place(&mut self.0);
    }
}
```